src/java.base/unix/classes/sun/nio/fs/UnixCopyFile.java

```diff
@@ -628,7 +713,14 @@ static void copy(final UnixPath source,
 
     // -- native methods --
 
-    static native void transfer(int dst, int src, long addressToPollForCancel)
+    // returns 0 on success, INTERRUPTED if the system call was interrupted,
+    // UNSUPPORTED_CASE if the call does not work with the supplied parameters,
+    // or UNSUPPORTED if the operation is not supported on this platform
+    static native int transfer0(int dst, int src, long addressToPollForCancel)
+        throws UnixException;
+
+    static native void copy0(int dst, int src, long address, int transferSize,
+                             long addressToPollForCancel)
```

Comment:
The naming is a bit unfortunate here. transferTo0 is copy with buffers in user-space. copy0 is with kernel support such as sendfile64 or fcopyfile. I think we need clearer names and improve the comments on the declarations here so that it's clear what they do. I think that's the first step before we split up the high-level copyFile method to make it easier to read and maintain.

Reply:
Refactored in 04d7502280df1dd0c2dccf34cc93450d86458e81.